lib/libc/gen/mips/longjmp.c

```diff
@@ -68,28 +67,29 @@ void longjmp(jmp_buf env, int val) {
   uc.uc_mcontext.__gregs[_REG_V0] = val;
 
   /* Copy saved registers */
-  uc.uc_mcontext.__gregs[_REG_S0] = sc->sc_regs[_R_S0];
-  uc.uc_mcontext.__gregs[_REG_S1] = sc->sc_regs[_R_S1];
-  uc.uc_mcontext.__gregs[_REG_S2] = sc->sc_regs[_R_S2];
-  uc.uc_mcontext.__gregs[_REG_S3] = sc->sc_regs[_R_S3];
-  uc.uc_mcontext.__gregs[_REG_S4] = sc->sc_regs[_R_S4];
-  uc.uc_mcontext.__gregs[_REG_S5] = sc->sc_regs[_R_S5];
-  uc.uc_mcontext.__gregs[_REG_S6] = sc->sc_regs[_R_S6];
-  uc.uc_mcontext.__gregs[_REG_S7] = sc->sc_regs[_R_S7];
-  uc.uc_mcontext.__gregs[_REG_S8] = sc->sc_regs[_R_S8];
+  uc.uc_mcontext.__gregs[_REG_S0] = sc_uc->uc_mcontext.__gregs[_REG_S0];
+  uc.uc_mcontext.__gregs[_REG_S1] = sc_uc->uc_mcontext.__gregs[_REG_S1];
+  uc.uc_mcontext.__gregs[_REG_S2] = sc_uc->uc_mcontext.__gregs[_REG_S2];
+  uc.uc_mcontext.__gregs[_REG_S3] = sc_uc->uc_mcontext.__gregs[_REG_S3];
+  uc.uc_mcontext.__gregs[_REG_S4] = sc_uc->uc_mcontext.__gregs[_REG_S4];
+  uc.uc_mcontext.__gregs[_REG_S5] = sc_uc->uc_mcontext.__gregs[_REG_S5];
+  uc.uc_mcontext.__gregs[_REG_S6] = sc_uc->uc_mcontext.__gregs[_REG_S6];
+  uc.uc_mcontext.__gregs[_REG_S7] = sc_uc->uc_mcontext.__gregs[_REG_S7];
+  uc.uc_mcontext.__gregs[_REG_S8] = sc_uc->uc_mcontext.__gregs[_REG_S8];
 #if !defined(__mips_abicalls)
-  uc.uc_mcontext.__gregs[_REG_GP] = sc->sc_regs[_R_GP];
+  uc.uc_mcontext.__gregs[_REG_GP] = sc_uc->uc_mcontext.__gregs[_REG_GP];
 #endif
-  uc.uc_mcontext.__gregs[_REG_SP] = sc->sc_regs[_R_SP];
-  uc.uc_mcontext.__gregs[_REG_RA] = sc->sc_regs[_R_RA];
-  uc.uc_mcontext.__gregs[_REG_EPC] = sc->sc_pc;
+  uc.uc_mcontext.__gregs[_REG_SP] = sc_uc->uc_mcontext.__gregs[_REG_SP];
+  uc.uc_mcontext.__gregs[_REG_RA] = sc_uc->uc_mcontext.__gregs[_REG_RA];
+  uc.uc_mcontext.__gregs[_REG_EPC] = sc_uc->uc_mcontext.__gregs[_REG_EPC];
 
   /* Copy FP state */
-  if (sc->sc_fpused) {
+  if (sc_uc->uc_flags & _UC_FPU) {
     /* FP saved regs are $f20 .. $f31 */
-    memcpy(&uc.uc_mcontext.__fpregs.__fp_r.__fp_regs[20], &sc->sc_fpregs[20],
-           32 - 20);
-    uc.uc_mcontext.__fpregs.__fp_csr = sc->sc_fpregs[_R_FSR - _FPBASE];
+    memcpy(&uc.uc_mcontext.__fpregs.__fp_r.__fp_regs[20],
+           &sc_uc->uc_mcontext.__fpregs.__fp_r.__fp_regs[20],
+           (32 - 20) * sizeof(float));
```

Comment:
I think it's better to use `sizeof(fpregister_t)` here.

Reply:
done